### PR TITLE
feat: update to `uglify-es`

### PR DESCRIPTION
**Pull request by Hartorn**
I tried this, because I wanted to be able to use webpack, with uglify es and not js.

Basically, I moved the options for Uglify into uglifyOptions key, and added an option for ES5 or ES6.

I also changed the code, to use the uglify.minify method.

Not sure if I understood something wrong, but I hope this can help. 
I can correct this if needed (tested on some of my projects, and it seemed to work)

EDIT : 
Here is a quick sum up of the modifications, please ask for precisions if needed.

 * Uglify now expose a single high level function `minify(code, options)`, which go through all the steps of minification (parse, compress, mangle, output)
 * it does not throw errors anymore, so the try-catch is not needed (test if there is an error key in the return of the function)
 * it is possible to get the warnings in the result also, so it is not needed anymore to mingle with AST_NODE `warn_function`
 * it is possible to give a filter for comments, so we can hook on that for exporting comments
 * it is possible to give a preamble in the output, so we can use that for the banner thing (License file, comments, ...)

Basically, we do not need to use any other function than minify from uglify, it is mainly about giving proper options, handle the source maps and output.

I did not had time to update the docs, will do in the next days.

### CODE REVIEW

**Comment by michael-ciniawsky** [comment]
Blocked by #56

File: src/index.js [hunk 1]
```diff
@@ -4,29 +4,31 @@
...
 "use strict";
 
-const SourceMapConsumer = require("source-map").SourceMapConsumer;
 const SourceMapSource = require("webpack-sources").SourceMapSource;
 const RawSource = require("webpack-sources").RawSource;
-const ConcatSource = require("webpack-sources").ConcatSource;
 const RequestShortener = require("webpack/lib/RequestShortener");
 const ModuleFilenameHelpers = require("webpack/lib/ModuleFilenameHelpers");
-const uglify = require("uglify-js");
+const uglifyEs = require("uglify-es");
```
Comment (michael-ciniawsky) [inline src/index.js]: Better dynamically required in L24?

File: src/index.js [hunk 2]
```diff
@@ -4,29 +4,31 @@
...
-		if(typeof options.compressor !== "undefined") options.compress = options.compressor;
-		this.options = options;
+
+		this.options = Object.assign({}, options);
+		this.options.test = this.options.test || /\.js($|\?)/i;
+		this.options.warningsFilter = this.options.warningsFilter || (() => true);
+
+		this.uglifyOptions = Object.assign({}, this.options.uglifyOptions || {}, { sourceMap: null });
+
+		this.uglify = this.options.es6 ? uglifyEs : uglifyJs;
```
Comment (michael-ciniawsky) [inline src/index.js]:
```
this.uglify = this.options.es6 ? require('uglify-es') : require('uglify-js');
```

File: src/index.js [hunk 3]
```diff
@@ -4,29 +4,31 @@
...
+const uglifyEs = require("uglify-es");
+const uglifyJs = require("uglify-js");
 
 class UglifyJsPlugin {
 	constructor(options) {
 		if(typeof options !== "object" || Array.isArray(options)) options = {};
-		if(typeof options.compressor !== "undefined") options.compress = options.compressor;
-		this.options = options;
+
+		this.options = Object.assign({}, options);
```
Comment (michael-ciniawsky) [inline src/index.js]:
```
this.options = options || {};
```

There is nothing to assign here (e.g `defaults`)

File: src/index.js [hunk 4]
```diff
@@ -36,10 +38,11 @@ class UglifyJsPlugin {
 				const files = [];
 				chunks.forEach((chunk) => files.push.apply(files, chunk.files));
 				files.push.apply(files, compilation.additionalChunkAssets);
-				const filteredFiles = files.filter(ModuleFilenameHelpers.matchObject.bind(undefined, options));
+				const filteredFiles = files.filter(ModuleFilenameHelpers.matchObject.bind(undefined, this.options));
 				filteredFiles.forEach((file) => {
-					const oldWarnFunction = uglify.AST_Node.warn_function;
-					const warnings = [];
+					// Copy uglify options
+					const uglifyOptions = Object.assign({}, this.uglifyOptions);
```
Comment (michael-ciniawsky) [inline src/index.js]:
```
const options = this.uglifyOptions || {};
```

File: src/index.js [hunk 5]
```diff
@@ -58,156 +61,30 @@ class UglifyJsPlugin {
...
-									banner = banner(commentsFile);
-								}
-								if(banner) {
-									outputSource = new ConcatSource(
-										"/*! " + banner + " */\n", outputSource
-									);
-								}
-							}
+						if(minifyResult.error) {
+							throw minifyResult.error;
```
Comment (michael-ciniawsky) [inline src/index.js]:
```
this.emitError(minifyResult.error)
```

? Not 💯 if `this.emitError` is Loader API only

**Comment by michael-ciniawsky** [comment]
Not sure about the semver tbh, guessing it is `major`

**Comment by graingert** [comment]
What's the point of supporting uglify-js doesn't uglify-es support a superset of the functionality of uglify-js?

**Comment by graingert** [comment]
No need for peerdep either, that was just a hack to support the harmony branch on github

**Comment by michael-ciniawsky** [comment]
> What's the point of supporting uglify-js doesn't uglify-es support a superset of the functionality of uglify-js?

Yeah... good point 😛

**Comment by Hartorn** [comment]
@michael-ciniawsky I will try to detail more soon, and to take in account your remarks.
@graingert I will change to dependencies then

I guess I was tired when reading uglify doc, I thought there were backward compatibility problems. 
Indeed, uglify-js is not needed

**Comment by graingert** [comment]
I imagine you'll need to update the documentation too.

**Comment by michael-ciniawsky** [comment]
> @michael-ciniawsky I will try to detail more soon, and to take in account your remarks.
> @graingert I will change to dependencies then

@Hartorn Yep this would be highly apprectiated, thx in advance

> I guess I was tired when reading uglify doc, I thought there were backward compatibility problems.
Indeed, uglify-js is not needed

All fine, this is a mess for weeks now :D, we need to backport the tests first to see what's maybe failing between `uglifyjs && uglifyes` then, it will make the transition a lot easier. Hopefully we can land PR (#56) for this soon :)

**Comment by Hartorn** [comment]
Okay, now I think the old options should work : 

Plugin options : 
* test
* include
* exclude 
* sourceMap
* warningsFilter
* extractComments
  * filename
  * banner
  * condition
* uglifyOptions

The uglifyOptions can take everything there : https://github.com/mishoo/UglifyJS2/tree/harmony#user-content-minify-options, except sourceMap part, which is given by the plugin.
I think it is better in a sub-object like this, would be a mess else.

**Comment by Hartorn** [comment]
@michael-ciniawsky @graingert Here is a quick sum up of the modifications, please ask for precisions if needed.

 * Uglify now expose a single high level function `minify(code, options)`, which go through all the steps of minification (parse, compress, mangle, output)
 * it does not throw errors anymore, so the try-catch is not needed (test if there is an error key in the return of the function)
 * it is possible to get the warnings in the result also, so it is not needed anymore to mingle with AST_NODE `warn_function`
 * it is possible to give a filter for comments, so we can hook on that for exporting comments
 * it is possible to give a preamble in the output, so we can use that for the banner thing (License file, comments, ...)

Basically, we do not need to use any other function than minify from uglify, it is mainly about giving proper options, handle the source maps and output.

I did not had time to update the docs, will do in the next days.

File: src/index.js [hunk 6]
```diff
@@ -10,172 +10,202 @@ const RawSource = require("webpack-sources").RawSource;
...
+					} else {
+						let outputSource = (minifyResult.map ?
+							new SourceMapSource(minifyResult.code, file, JSON.parse(minifyResult.map), input, inputSourceMap) :
+							new RawSource(minifyResult.code));
+
+						asset.__UglifyJsPlugin = compilation.assets[file] = outputSource;
+
+						// Handling warnings
+						if(minifyResult.warnings) {
+							const buildWarnings = this.buildWarnings(minifyResult.warnings, file, sourceMap, this.options.warningsFilter, requestShortener);
```
Comment (michael-ciniawsky) [inline src/index.js]: `const buildWarnings => const warnings`

File: src/index.js [hunk 7]
```diff
@@ -10,172 +10,202 @@ const RawSource = require("webpack-sources").RawSource;
...
-						if(options.sourceMap) {
-							map = uglify.SourceMap({ // eslint-disable-line new-cap
-								file: file,
-								root: ""
-							});
-							output.source_map = map; // eslint-disable-line camelcase
+					}
+
+					// Calling uglify
+					let minifyResult = uglify.minify({ [file]: input }, uglifyOptions);
```
Comment (michael-ciniawsky) [inline src/index.js]: `let minifyResult => let result`

**Comment by michael-ciniawsky** [comment]
@Hartorn Could you change the base branch to `feature-defaults` please ? :)

**Comment by Hartorn** [comment]
@michael-ciniawsky I changed the base, rebased my branch and corrected the lint according to the new rules (much better in ES6 :) )

**Comment by michael-ciniawsky** [comment]
@Hartorn thx

**Comment by michael-ciniawsky** [comment]
Sry it landed in master now 😅 , could you switch again and rebase against current `master` one more time please 😛

File: package.json [hunk 8]
```diff
@@ -82,4 +81,4 @@
   },
   "homepage": "https://github.com/webpack-contrib/uglifyjs-webpack-plugin",
   "license": "MIT"
-}
+}
```
Comment (graingert) [inline package.json]: invalid text file. Files a sequence lines terminated with a newline character.

**Comment by hulkish** [comment]
@Hartorn can you please rebase to master?

**Comment by michael-ciniawsky** [comment]
Please rebase against current master, sry for the inconvenience again, this is definitely the next to land 💯 😛

**Comment by hulkish** [comment]
@Hartorn Also, i noticed you removed the `try`/`catch` in the loop for uglifyjs. Can you explain that a bit?

**Comment by Hartorn** [comment]
@hulkish @michael-ciniawsky done rebasing. I will have a look at the test now.

@hulkish Uglify does not throw error anymore, instead it is given in the result object.
See the few lines above https://github.com/mishoo/UglifyJS2/tree/harmony#minify-options

**Comment by michael-ciniawsky** [comment]
⚠️  The git commit author history seems to be messed up, you either need to edit the authors via git or open a fresh PR, otherwise the CLA Bot will complain and block the PR

**Comment by hulkish** [comment]
@Hartorn I recommend making a new pr based on master and just re apply your changes - sorry for the hassle, but I think this is your easiest route tbh.

**Comment by hulkish** [comment]
@Hartorn Nvm, i see you've already included your updates to the tests: https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/55/files#diff-3fbea51fb89113c24da0c56675865113R30
